Static Init, and some new functions

Sanae6 opened the pull request:
This PR introduces static LValue init for Reals, Strings, and Arrays. Along with it comes the functions draw_set_circle_precision, object_get_name, and room_get_viewport. Views aren't implemented in rendering yet, so I'm not going further with room calls until that is complete.

Sanae6 committed:
Many changes made, summarized changelist fully in commit message
Changed Data.IFF to use a path provided by program arguments
Moved the File.Exists check from Data.IFF to Program.Main
Automatically adds OpenTK.Key keys to the InputMapping in Runner.Input
Added functions draw_set_circle_precision and room_get_name
draw_circle now uses set precision value to calculate its' vertex positions
An assumed to be successful attempt was made to implement PushBuiltin instruction (used for a test case with room_get_name(room))
Many unnecessary this accesses and LValue.Value to double or string have been cleaned up and converted to their proper equivalents
LValue double comparison now does NearlyEqual comparisons with room for error, see comment for StackOverflow link to original code
CurrentRoom variable added to Game to store the current room's corresponding object (used in built-in variable "room")

Sanae6 committed:
Merge branch 'master' of https://github.com/nommiin/Luna

 Conflicts:
	Luna/Assets/LCode.cs
	Luna/Runner/Function.cs
	Luna/Runner/Instruction.cs
	Luna/Types/LInstance.cs
	Luna/Types/LValue.cs

Sanae6 committed:
Fixing LValue from merge for Function.Handler in Instruction

nommiin committed:
Rewrote array support, removed some comments+debug messages, changed some variables named "_inst" to "_instruction" to avoid confusion with LInstances

Sanae6 committed:
Fixed another mistake with Functions and LValue not being returned
Set room variable before step like planned
Fixed more things but you pushed so i pull

nommiin committed:
Added exception to setowner processing for array literals, added array_create support

nommiin committed:
Removed Sample/ folder from gitignore, pushed sample WADs and projects, added support for pushing built-in variables, added CurrentRoom variable to Game class, added room_get_name function

nommiin committed:
Fixed typo where yprevious variable was being set to the y variable (thanks Shana6), made rudimentary code style guide

Sanae6 committed:
Add Samples as Solution Folder

Sanae6 committed:
Use args instead of hardcoded string to choose WAD

Sanae6 committed:
Static initializers for LValues to speed up function creation

Sanae6 committed:
Some view variables, room_get_viewport, object_get_name and a todo comment

Sanae6 committed:
draw_set_circle_precision and Game.CirclePrecision, along with some LValue static inits

Sanae6 committed:
unintelligible character spam

Sanae6 committed:
oops :P

nommiin commented:
Looks great! There's some places where code style isn't enforced but those can be either fixed pre-pull or after, either works. The static initalizers for LValue will definitely help clean things up!

Some additional notes:
- Be sure not to use the "ReadBoolean" method of BinaryReader, as the asset compiler writes 0 or 1 as an Int32 for booleans. I believe ReadBoolean will only read 1 byte and will become incorrectly offset for further values. I could look into adding an extension method for BinaryReader to read these Int32s instead of having to repeatedly use ternaries
- Ideally we'll use LColour for any colours with OpenTK, maybe there can be a method for LColour which will return Color4?
- Unless I missed it a fix for it, `object_get_name`'s method is named `room_get_height` 

Otherwise, this pull is great! Thanks again for contributing!

Sanae6 commented:
First, will keep boolean issue in mind for future parsing.
Second, LColour totally slipped my mind, I'll use that instead.
Third, yeah I saw that too, it was fixed before you mentioned it in a local commit 👍

Sanae6 committed:
thank god for circle precision
rendering is stretched and i'm not sure why, so i'm looking into that when possible

Sanae6 committed:
Merge branch 'master' of https://github.com/nommiin/Luna

 Conflicts:
	Luna/Assets/LCode.cs
	Luna/Assets/LObject.cs
	Luna/Assets/LRoom.cs
	Luna/Assets/LSprite.cs
	Luna/Data/ChunkHandler.cs
	Luna/Program.cs
	Luna/Runner/Domain.cs
	Luna/Runner/Function.cs
	Luna/Runner/Game.cs
	Luna/Runner/Input.cs
	Luna/Runner/Instruction.cs
	Luna/Types/LEvent.cs
	STYLE.md
	Sample/LunaTest/objects/Object1/Create_0.gml

Sanae6 committed:
Merging fails, day 12

Sanae6 committed:
Merge branch 'statinits' of https://github.com/Shana6/Luna

 Conflicts:
	Luna/Runner/Function.cs

Sanae6 committed:
idk lol more merging mistakes?

Sanae6 committed:
Merge branch 'master' of https://github.com/nommiin/Luna into statinits

 Conflicts:
	Luna/Program.cs

Sanae6 committed:
okay and circles written again quickly and fixed

Sanae6 committed:
Region off the object_get_name
